draft LMS preview doc (#17895)
cockroachdb/docs: src/current/v23.2/live-migration-service.md

MOLT LMS is self-hosted on [Kubernetes](https://kubernetes.io/) and [configured using Helm](#configuration). At a high level, the LMS consists of the following:

- A number of proxy [instances](#lms-instances) (running in separate Kubernetes pods) across which application traffic is distributed and routed to the source and target databases.
- An "orchestrator" service (running in a single Kubernetes pod) that coordinates the proxy instances and sends the cutover commands.

This page describes how to [install](#installation), [configure](#configuration), [secure](#security), and [use the LMS](#molt-lms-cli) to perform a live migration. {% comment %}For more information, see [Migration Strategy: Live Migration]({% link {{ page.version.version }}/migration-strategy-live-migration.md %}).{% endcomment %}

## Terminology

- A *live migration* involves maintaining two production databases (a source and target database) and replicating data between them until a final cutover.
- The *source of truth* is the database that serves reads and writes to the application during a live migration. A cutover switches the source of truth.

> - *Shadowing* is the replication of writes from the source of truth to the target database. The LMS supports multiple [shadowing modes](#shadowing-modes).

Reviewer: i would say it is "the replaying of traffic on the target database", as reads get routed through there.

Author: Noted. Can I still use "replication" instead, or do you have a specific preference for "replaying"?

Reviewer: "replay" is better for me - we're not replicating data, we're executing the same SQL traffic patterns

## Requirements

- [Kubernetes](https://kubernetes.io/) cluster
- [Helm](https://helm.sh/docs/intro/install/) package manager for Kubernetes

#### Supported databases

> - [PostgreSQL]({% link {{ page.version.version }}/migrate-from-postgres.md %}) (source)

Reviewer 2: We also support CRDB as a source.

- [MySQL]({% link {{ page.version.version }}/migrate-from-mysql.md %}) (source)
- CockroachDB (target)

## Installation

The LMS definitions are located in the [`molt` repository](https://molt.cockroachdb.com/):

- [`molt-lms` Helm chart](https://molt.cockroachdb.com/charts/lms/Chart.yaml)

> - [`molt-lms` values file](https://molt.cockroachdb.com/charts/lms/values.yaml) for [configuration](#configuration)

Reviewer 2: I'm not sure if this is the right thing to link to unless we are creating redirect with this url to the actual github repo. The molt.cockroachdb.com item is a packaged chart not like what you would view in github directly. I think this maybe more useful to point to the exact github file. cc @ryanluu12345

Author: OK, I had the direct GitHub links originally. I'm not clear on what exactly will be at `molt.cockroachdb.com` and where.

Reviewer 3: So at `molt.cockcroachdb.com/lms/cli/` will live all the binaries. At `molt.cockroachdb.com/lms/charts` will live the chart that someone can use to `helm repo add`. But that will be the fully packaged helm chart and not the raw `.yaml` themselves. @Jeremyyang920 is correct that the raw charts will live in the Github itself.

When you install the `molt-lms` chart, the LMS proxy instances and orchestrator are initialized as Kubernetes pods:

{% include_cached copy-clipboard.html %}
~~~ shell
kubectl get pods
~~~

~~~
NAME                                          READY   STATUS      RESTARTS      AGE
lms-molt-lms-5cbd7c748-bzbgn                  1/1     Running     0             2m30s
lms-molt-lms-5cbd7c748-g5zqf                  1/1     Running     0             2m34s
lms-molt-lms-5cbd7c748-lbjjz                  1/1     Running     0             2m3
lms-molt-lms-orchestrator-596d4b54d8-c4v97    1/1     Running     0             21m
...
~~~

You will see `lms-molt-lms` pods that match the configured [number of LMS instances](#lms-instances), along with one `lms-molt-lms-orchestrator` pod.

## Configuration

> To configure the LMS, override the values defined in [`values.yaml`](https://molt.cockroachdb.com/charts/lms/values.yaml). For information on setting Helm chart values, see the [Helm documentation](https://helm.sh/docs/helm/helm_upgrade/).

Reviewer 3: As noted above, the values themselves will be in the YAML but at `molt.cockroachdb.com/lms/charts` (note that the paths have changed to be lms prefixed first since we have different binaries we are sharing inside of lms) the `.tgz` package and `index.yaml` will be.

Author: > the .tgz package and index.yaml will be

I'm confused on what these are. Can you elaborate?

Reviewer 3: Yep of course, so when you do a `helm package` command, it will generate a `.tgz` zipped archive for all the YAMLs in the chart. The `index.yaml` is a manifest basically that describes the summary details of the zipped archive. Those are the items that live in the bucket  at `molt.cockroachdb.com/lms/charts`. These are just implementation details of how the chart gets published.

Reviewer 3: The source of the charts lives on Github. That's all you need to know

{{site.data.alerts.callout_info}}
Values that are specified in both `lms` and `orchestrator` must have the identical setting.
{{site.data.alerts.end}}

This section describes the most important and commonly used values.

#### Source dialect

~~~ yaml
lms:
  sourceDialect: ""
...
orchestrator:
  sourceDialect: ""
~~~

You **must** provide a string value for `sourceDialect`, which specifies the dialect of your source database. Supported dialects are:

- `postgres`: PostgreSQL
- `mysql`: MySQL

> - `crdb`: CockroachDB

Author: Just checking - this is case-sensitive?

> - `crdb`: CockroachDB

Reviewer 3: Based on the source code, I think we accept `cockroach` and `crdb` to specify the dialect, since we use the dialect flag. I could be mistaken though. @Jeremyyang920 to clarify:
```
const (
	Cockroach Dialect = iota + 1
	MySQL
	PostgreSQL
	CockroachSource
	CockroachTarget
)

var DialectStringRepresentations = map[Dialect][]string{
	Cockroach:       {"crdb", "cockroach"},
	CockroachSource: {"crdb_src"},
	CockroachTarget: {"crdb_dst"},
	MySQL:           {"mysql"},
	PostgreSQL:      {"pg", "postgres", "postgresql", "libpq", "pgsql", "pq", "psql"},
}
```

> - `crdb`: CockroachDB

Reviewer: not case sensitive, we take either.

  }
~~~

For more details on storing the LMS configuration in an external secret, see [Manage external secret](#manage-external-secret).

#### Certificates

~~~ yaml
lms:
  sslVolumes: {}
  sslVolumeMounts: {}
...
orchestrator:
  sslVolumes: {}
  sslVolumeMounts: {}
~~~

`sslVolumes` and `sslVolumeMounts` specify [volumes](https://kubernetes.io/docs/concepts/storage/volumes/#secret) and mount paths that contain server-side certificates for the LMS instances and orchestrator. For security recommendations, see [Security](#security).

<a name="cert-var"></a>

When you specify a volume in `sslVolumes` and `sslVolumeMounts`, you must also [set its corresponding config key](#manage-external-secret) to the path specified in `sslVolumeMounts.mountPath`.

|      Key      |                        Description                         |
|---------------|------------------------------------------------------------|
| `SSL_CA`      | Mount path to the SSL CA certificate for the LMS.          |
| `SSL_CERT`    | Mount path to the SSL certificate for the LMS.             |
| `SSL_KEY`     | Mount path to the SSL key for the LMS.                     |

> | `CA_TLS_CERT` | Mount path to the TLS CA certificate for the orchestrator. |

Reviewer 3: We may want to specify that this is *optional* and mainly used if you self-sign your certs. Otherwise, you can use a root CA like Let's Encrypt or AWS or etc.

Author: Does this also go for `SSL_CA`?

> | `TLS_CERT`    | Mount path to the TLS certificate for the orchestrator.    |

Reviewer 3: For these, might want to note that they are the server side certs

Reviewer 3: Scratch that, you mention it in the intro.

#### Manage external secret

Use an external secrets manager such as [External Secrets Operator](https://external-secrets.io/latest/) to create and manage a Kubernetes secret containing your LMS configuration. For information on Kubernetes secrets, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/secret/).

In the [LMS configuration](#configuration), specify the name of your external secret with `configSecretName`:

~~~ yaml
lms:
  configSecretName: "secret-name"
...
orchestrator:
  configSecretName: "secret-name"
~~~

WIthin the secret, set LMS config keys inside a JSON object named `config.json`. For example, in an `ExternalSecret` named `lms-config` created with External Secrets Operator:

~~~ yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: lms-config
spec:
  ...
  target:
    name: lms-config
    ...
      data:
        config.json: |
          {

>             "INIT_SOURCE": "mysql://{username}:{password}@{protocol}({host}:{port})/{database}?sslmode=verify-full?sslrootcert=path/to/mysql.ca&sslcert=path/to/mysql.crt&sslkey=path/to/mysql.key",

Reviewer 3: I recommend showing two blocks here: 
1. Parametrized version that index properties from the external secret that is pulled (how customers should do it. (i.e. `"SSL_CA": lms-config.ca). Or however the parametrized version looks.
2. The example you have that shows the hardcoded values.

Reviewer 2: Specifically for the critical keys like init_source and init_target. 
The paths to the files being hard coded im not as concerned about. Also they need the file name at the end not just `/app/certs`

    "INIT_TARGET": "postgresql://{username}:{password}@{host}:{port}/{database}?sslmode=verify-full?sslrootcert=path/to/ca.crt&sslcert=path/to/client.username.crt&sslkey=path/to/client.username.key"
  }
~~~

#### Configure orchestrator and client certificates

Mount the orchestrator certificate, key, and (optional) CA certificate to separate volumes. Cockroach Labs recommends mounting certificates to `/app/certs`.

For example:

~~~ yaml
orchestrator:
  sslVolumes:
    - name: orch-ca
      secret:
        secretName: orch-ca
    - name: orch-cert
      secret:
        secretName: orch-cert
    - name: orch-key
      secret:
        secretName: orch-key
  sslVolumeMounts:
    - mountPath: "/app/certs"
      name: orch-ca
      readOnly: true
    - mountPath: "/app/certs"
      name: orch-cert
      readOnly: true
    - mountPath: "/app/certs"
      name: orch-key
      readOnly: true
~~~

Set the [corresponding config keys](#cert-var) to the same mount path. For example, within `config.json` in the [external secret](#manage-external-secret):

~~~ yaml
config.json: |
  {
    "CA_TLS_CERT": "/app/certs",
    "TLS_CERT": "/app/certs",
    "TLS_KEY": "/app/certs",
  }
~~~

> This also requires that you create and specify a CLI client certificate, key, and (optional) CA certificate. It's easiest to specify these with the following environment variables:

Reviewer 3: Mention how this will be in the shell where you're running the CLI

To install `molt-lms-cli`, [download the binary](https://molt.cockroachdb.com/).

### Commands

|       Command        |                                                                                     Usage                                                                                      |
|----------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `initialize`         | Set up the required objects for running the LMS. You must run this before using the LMS.                                                                                       |
| `connections list`   | List all client connections to the LMS and their most recent queries.                                                                                                          |
| `cutover consistent` | Specify a [consistent cutover](#consistent-cutover). You must also specify `begin`, `commit`, or `abort`. For usage details, see [Consistent cutover](#consistent-cutover).    |

> | `begin`              | Begin a consistent cutover. This pauses traffic to the source of truth.                                                                                                        |

Reviewer 4: ```suggestion
| `begin`              | Begin a consistent cutover. This pauses traffic to the source database.                                                                                                       |
```

> | `commit`             | Commit a consistent cutover. This switches the source of truth to the target database.                                                                                         |

Reviewer 4: ```suggestion
| `commit`             | Commit a consistent cutover. This starts traffic to the target database after running. This command is only effective when run after `begin`                                                                                        |
```

Author: How does this look?

```
Commit a consistent cutover. This resumes traffic on the target database, which becomes the source of truth. This is only effective after running `cutover consistent begin`.
```

> | `abort`              | Abort a consistent cutover after running `consistent cutover begin`, unless you have also run `consistent cutover commit`. This resumes traffic to the source of truth.        |

Reviewer 4: ```suggestion
| `abort`              | Abort a consistent cutover after running `consistent cutover begin`, unless you have also run `consistent cutover commit`. This resumes traffic to the source database.        |
```

> | `cutover immediate`  | Initiate an [immediate cutover](#immediate-cutover). This switches the source of truth to the target database. For usage details, see [Immediate cutover](#immediate-cutover). |

Reviewer 3: @cotedm @ZhouXing19 @taroface this is currently a `TODO` command. Should we remove this from both the binary and the docs?

Reviewer 4: Yes, if it is a `TODO` let's remove it. @taroface this will also mean hiding the section on Immediate cutover for now until we have this in the CLI.

Author: Will this be a `TODO` until Roachfest?

Reviewer 5: Yes immediate cutover won't be ready on Roachfest.

| `status`             | Display the current configuration of the LMS instances.                                                                                                                        |

### Flags

|         Flag         |                                                                                                    Description                                                                                                    |
|----------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `--orchestrator-url` | The URL for the orchestrator, using the [configured port](#service-type). This flag is required unless the value is exported as an environment variable using `export CLI_ORCHESTRATOR_URL="{orchestrator-URL}"`. |

Reviewer 3: I think it's also worth mentioning that if someone is using SSL certs, their protocol prefix is: `https` versus `http`

| `--tls-ca-cert`      | The path to the TLS CA certificate. This can also be [exported](#configure-orchestrator-and-client-certificates) as an environment variable using `export CLI_TLS_CA_CERT="{path-to-cli-ca-cert}"`.               |
| `--tls-client-cert`  | The path to the TLS client certificate. This can also be [exported](#configure-orchestrator-and-client-certificates) as an environment variable using `export "{path-to-cli-client-cert}"`.                       |
| `--tls-client-key`   | The path to the TLS client key. This can also be [exported](#configure-orchestrator-and-client-certificates) as an environment variable using `export "{path-to-cli-client-key}"`.                                |

> ## Shadowing modes

Reviewer 3: Deferring to @otan and @Jeremyyang920 for this logic.

### `none`

<img src="{{ 'images/v23.2/migrations/lms_none.svg' | relative_url }}" alt="MOLT LMS shadowing mode - none" />

`shadowMode: none` disables shadowing.

- The LMS sends application requests to the source of truth only.
- Results from the source of truth are returned to the application.
- Writes must be manually replicated from the source database to the target database.

You can use this mode to perform a [consistent cutover](#consistent-cutover), along with another tool such as [change data capture (CDC)]({% link {{ page.version.version }}/cdc-queries.md %}) that replicates writes to the target database. {% comment %}For an example, see [Consistent cutover without shadowing](#consistent-cutover-without-shadowing).{% endcomment %}

### `async`

<img src="{{ 'images/v23.2/migrations/lms_async.svg' | relative_url }}" alt="MOLT LMS shadowing mode - async" />

`shadowMode: async` writes to both databases.

- The LMS sends application requests to the source of truth and target database in asynchronous threads, and waits only for the source of truth to respond.

> - Results from the source of truth are returned to the application.

Reviewer: can you add a note here: "if an asynchronous request has not yet completed, subsequent asynchronous requests will be dropped"

Author: Added. But how/when do async requests start up again? Only when the last async request is completed?

Reviewer: yeah, more or less. i kinda think we may want to avoid documenting / suggesting it as a whole, cc @cotedm wdyt?

Author: per @cotedm we'll add a callout clarifying that async shadowing mode is for testing.

You can use this mode to confirm that your queries succeed on CockroachDB without verifying performance or correctness.

### `sync`

<img src="{{ 'images/v23.2/migrations/lms_sync.svg' | relative_url }}" alt="MOLT LMS shadowing mode - sync" />

`shadowMode: sync` writes to both databases.

- The LMS sends application requests to the source of truth and the target database, and waits for each to respond.
- Results from the source of truth are returned to the application.
- Results from the non-source of truth are discarded.

You can use this mode to perform an [immediate cutover](#immediate-cutover).

### `strict-sync`

<img src="{{ 'images/v23.2/migrations/lms_strict-sync.svg' | relative_url }}" alt="MOLT LMS shadowing mode - strict-sync" />

`shadowMode: strict-sync` writes to both databases and enforces correctness on both databases.

- The LMS sends application requests to the source of truth and the target database, and waits for each to respond.
- Results from the source of truth are returned to the application.

> - If the non-source of truth returns an error, that error is returned instead of the result from the source of truth.

Reviewer: if the source of truth errors as well, the non-source-of-truth will return a non-error.

You can use this mode to perform an [immediate cutover](#immediate-cutover).

## Perform a cutover

### Consistent cutover

> A consistent cutover maintains data consistency with [minimal downtime]({% link {{ page.version.version }}/migration-overview.md %}#minimal-downtime). When using the LMS, consistent cutover is handled using the [`molt-lms-cli`](#molt-lms-cli) commands `cutover consistent begin` and `cutover consistent commit`, between which downtime occurs. The amount of downtime depends on the maximum duration of any transactions and queries that need to complete, and the time it takes for replication to catch up from the source to the target database.

Reviewer 4: ```suggestion
A consistent cutover maintains data consistency with [minimal downtime]({% link {{ page.version.version }}/migration-overview.md %}#minimal-downtime). When using the LMS, consistent cutover is handled using the [`molt-lms-cli`](#molt-lms-cli) commands `cutover consistent begin` and `cutover consistent commit`, during which application requests are queued and will be responded to after cutover. This delay in response time is related to the maximum duration of any transactions and queries that need to complete, and the time it takes for replication to catch up from the source to the target database. This is typically 1-2 minutes in a production system.
```
I want to get really specific in this section about what downtime means. It's not really an "outage" but rather a delay in responses. Everything still gets processed as long as requests don't time out so we want to be specific.

Author: Noted and updated!

{% comment %}
For more information about the consistent cutover approach, see [Migration Strategy: Live Migration]({% link {{ page.version.version }}/migration-strategy-live-migration.md %}).
{% endcomment %}

To perform a consistent cutover with the LMS:

{{site.data.alerts.callout_info}}
These steps assume you have already followed the overall steps to [prepare for migration]({% link {{ page.version.version }}/migration-overview.md %}#prepare-for-migration). In particular, [update your schema and application queries]({% link {{ page.version.version }}/migration-overview.md %}#update-the-schema-and-queries) to work with CockroachDB.
{{site.data.alerts.end}}

> 1. [Configure the LMS](#configuration) with your deployment details, and follow our [security recommendations](#security).

Reviewer 5: cc @otan @cotedm do we need to include technical details, especially the `_crdb_proxy_settings` table?

Reviewer 4: I'd prefer not to document this in the docs website. However, having it in the knowledge base for TSE or in a runbook internally I think is valuable. In general, customers shouldn't be messing with these internal details and if it's needed for troubleshooting, we should be holding their hand IMO

Reviewer 5: Ack, will document it in the internal runbook.

1. Set the shadowing mode to [`none`](#none).

	{{site.data.alerts.callout_danger}}
	Do not use the [`sync`](#sync) or [`strict-sync`](#strict-sync) shadowing modes when performing a consistent cutover. Data correctness and consistency cannot be guaranteed in these configurations.
	{{site.data.alerts.end}}

1. Set up ongoing replication between the source database and CockroachDB, using a tool such as [change data capture (CDC)]({% link {{ page.version.version }}/cdc-queries.md %}).

1. Send application requests to the LMS, which routes the traffic to the source database. The source database is designated the source of truth.

1. Use [MOLT Verify]({% link {{ page.version.version }}/molt-verify.md %}) to validate that the replicated data on CockroachDB is consistent with the source of truth.

> 1. Begin the consistent cutover. **This begins downtime**:

Reviewer 4: ```suggestion
1. Begin the consistent cutover. **Requests are now queued in LMS**:
```

Reviewer 5: Do you think we should specify the requests being queued here include both queries from the existing connections, and also new connection request towards the LMS instance?

Author: Added!

	{% include_cached copy-clipboard.html %}
	~~~ shell
	molt-lms-cli cutover consistent begin {flags}
	~~~

	This command tells the LMS to pause all application traffic to the source of truth. The LMS then waits for transactions to complete and prepared statements to close.

1. Verify that replication on CockroachDB has caught up with the source of truth. For example, write a marker record to the source database and check that it exists on CockroachDB.

	If you have an implementation that replicates back to the source database, this should be enabled before committing the cutover.

1. Once all writes have been replicated to the target database, commit the consistent cutover:

	{% include_cached copy-clipboard.html %}
	~~~ shell

> 	molt-lms-cli cutover consistent commit {flags}

Reviewer 5: for cutover consistent commit, we don't take flags at this moment.

Author: @ZhouXing19 Does this comment still stand? I thought you'd still need to pass the URL flag and TLS flags if necessary.

Reviewer 5: Yeah I remembered it wrong -- it does take flags, so I resolved this comment. Sorry for the noise!

	~~~

> 	This command tells the LMS to switch the source of truth to CockroachDB. Application traffic resumes on CockroachDB, and this ends downtime.

Reviewer 4: ```suggestion
	This command tells the LMS to switch the source of truth to CockroachDB. Application traffic resumes on CockroachDB, and requests are processed from the queue in LMS.
```

Reviewer 5: To be aligned with the `source database` previously, maybe we should say `switch the source of truth to the target database. Application traffic are now routed to the target database`

## Terminology

- A *live migration* keeps two production databases online (a source and a target database) and uses either replication or dual writing to keep data equal between them until a final cutover.
- The *source of truth* is the database that serves reads and writes to the application during a live migration. A cutover switches the source of truth.
- *Shadowing* is the replication of traffic on the target database. The LMS supports multiple [shadowing modes](#shadowing-modes).

## Requirements

- [Kubernetes](https://kubernetes.io/) cluster
- [Helm](https://helm.sh/docs/intro/install/) package manager for Kubernetes

#### Supported databases

- [PostgreSQL]({% link {{ page.version.version }}/migrate-from-postgres.md %}) (source)
- [MySQL]({% link {{ page.version.version }}/migrate-from-mysql.md %}) (source)
- CockroachDB (source and target)

## Installation

> To install the LMS, add the [Helm chart](https://molt.cockroachdb.com/lms/charts).

Reviewer 4: We'll need the direct link here: https://molt.cockroachdb.com/lms/charts/index.yaml

When you install the `molt-lms` chart, the LMS proxy instances and orchestrator are initialized as Kubernetes pods:

{% include_cached copy-clipboard.html %}
~~~ shell
kubectl get pods
~~~

~~~
NAME                                          READY   STATUS      RESTARTS      AGE
lms-molt-lms-5cbd7c748-bzbgn                  1/1     Running     0             2m30s
lms-molt-lms-5cbd7c748-g5zqf                  1/1     Running     0             2m34s
lms-molt-lms-5cbd7c748-lbjjz                  1/1     Running     0             2m3
lms-molt-lms-orchestrator-596d4b54d8-c4v97    1/1     Running     0             21m
...
~~~

> You will see `lms-molt-lms` pods that match the configured [number of LMS instances](#lms-instances), along with one `lms-molt-lms-orchestrator` pod.

Reviewer 2: This is just a bit of a clarification thing, but the pod base name is just `molt-lms` and `molt-lms-orchestrator` the `lms` prefix at the front depends on what you named your release as that can vary. So might be worth calling that out.

Author: Noted. Which param name corresponds to this prefix?

Reviewer 2: There isin't exactly field that sets the prefix.
If you set name override, that overrides the `molt-lms` or `molt-lms-orchestrator` part. So if I set name override to ryan. It would be `lms-ryan` and `lms-ryan-orchestrator`

If you set the full name override to say blah, it would be `blah` and blah-orchestrator`. I would just perhaps call out that in your example, our release name is `lms`.

## Configuration

To configure the LMS, override the values defined in the [Helm chart](#installation). For information on setting Helm chart values, see the [Helm documentation](https://helm.sh/docs/helm/helm_upgrade/).

{{site.data.alerts.callout_info}}
Values that are specified in both `lms` and `orchestrator` must have the identical setting.
{{site.data.alerts.end}}

This section describes the most important and commonly used values.

#### Source dialect

~~~ yaml
lms:
  sourceDialect: ""
...
orchestrator:
  sourceDialect: ""
~~~

You **must** provide a string value for `sourceDialect`, which specifies the dialect of your source database. Supported dialects are:

- `postgres`: PostgreSQL
- `mysql`: MySQL
- `cockroach`: CockroachDB

#### Shadowing

~~~ yaml
lms:
  shadowMode: none
~~~

`lms.shadowMode` specifies the shadowing behavior used by the LMS. This should depend on your specific migration requirements. For details, see [Shadowing modes](#shadowing-modes).

#### LMS instances

~~~ yaml
lms:
  replicaCount: 3
~~~

> `lms.replicaCount` determines the number of LMS instances created as `lms-molt-lms` pods on the Kubernetes cluster, across which application traffic is distributed. This defaults to `3`.

Reviewer 2: Similar to the previous comment, but `lms-molt-lms` may not be their pod name if they named the release something different or did a name override.

#### Source and target connections

The connections to the source database and CockroachDB are defined with the configuration keys:

- `INIT_SOURCE`: Connection string for the source database.
- `INIT_TARGET`: Connection string for a CockroachDB database.

Include client authentication details in the connection strings. For more details, see [Configure source and target certificates](#configure-source-and-target-certificates).

{{site.data.alerts.callout_success}}
For details about writing a CockroachDB connection string, see [Connect using a URL]({% link {{ page.version.version }}/connection-parameters.md %}#connect-using-a-url).
{{site.data.alerts.end}}

You should specify the config keys in an external Kubernetes secret and inside a JSON object. The JSON **must** be named `config.json`.

For example:

~~~ json
config.json: |
  {

>     "INIT_SOURCE": "mysql://{username}:{password}@{protocol}({host}:{port})/{database}?sslmode=verify-full?sslrootcert=path/to/mysql.ca&sslcert=path/to/mysql.crt&sslkey=path/to/mysql.key",

Reviewer 2: We can omit the protocol part and use conn strings for mysql that look like. 
mysql://admin:admin@127.0.0.1:3306/defaultdb. They can basically look like a crdb conn string.

## Security

{{site.data.alerts.callout_danger}}
Cockroach Labs **strongly** recommends the following:

- Manage your LMS configuration in an [external Kubernetes secret](#manage-external-secret).
- To establish secure connections between the LMS pods and with your client, set up TLS certificates for the [LMS](#configure-lms-certificate), [source database and CockroachDB](#configure-source-and-target-certificates), and [orchestrator and client](#configure-orchestrator-and-client-certificates).
{{site.data.alerts.end}}

#### Manage external secret

Use an external secrets manager such as [External Secrets Operator](https://external-secrets.io/latest/) to create and manage a Kubernetes secret containing your LMS configuration. For information on Kubernetes secrets, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/secret/).

> In the [LMS configuration](#configuration), specify the name of your external secret with `configSecretName`:

Reviewer 3: You might want to add a note that this will map 1:1 with the `target name` you specify in the external secret?

You also want to mention that these will be two distinct config secret names because the values that the LMS and orchestrator use are different.

~~~ yaml
lms:
  configSecretName: "secret-name"
...
orchestrator:
  configSecretName: "secret-name"
~~~

> WIthin the secret, set LMS config keys inside a JSON object named `config.json`. For example, in an `ExternalSecret` named `lms-config` created with External Secrets Operator:

Reviewer 2: ```suggestion
Within the secret, set LMS config keys inside a JSON object named `config.json`. For example, in an `ExternalSecret` named `lms-config` created with External Secrets Operator:
```

Reviewer 2: Should also call out the below is an example of using AWS SSM as your backend secret store.

~~~ yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: lms-config
spec:
  ...
  target:
    name: lms-config
    ...
      data:

>         config.json: |

Reviewer 3: Change requested here.

One thought I had is that there should be a separate external secret for:
1. `lms-config`
2. `orchestrator-config`

They are actually distinct because they take different config values. There are values that LMS cares about that the orchestrator has no notion of. Vice versa too. This diff has an example of which ones are relevant to each service: https://github.com/cockroachdb/molt/compare/main...user/rluu/run-lms-secrets-and-tls

            "TLS_CERT": "/app/certs",
            "TLS_KEY": "/app/certs",
            ...
          }
...
~~~

The preceding example defines config keys for the [source and target database connections](#source-and-target-connections) and mount paths for various [certificates](#certificates).

#### Configure LMS certificate

Mount the LMS certificate, key, and (optional) CA certificate to separate volumes. Cockroach Labs recommends mounting certificates to `/app/certs`.

For example:

~~~ yaml
lms:
  sslVolumes:
    - name: lms-ca
      secret:
        secretName: lms-ca
    - name: lms-cert
      secret:
        secretName: lms-cert
    - name: lms-key
      secret:
        secretName: lms-key
  sslVolumeMounts:
    - mountPath: "/app/certs"
      name: lms-ca
      readOnly: true
    - mountPath: "/app/certs"
      name: lms-cert
      readOnly: true
    - mountPath: "/app/certs"
      name: lms-key
      readOnly: true
~~~

Set the [corresponding config keys](#cert-var) to the same mount path. For example, within `config.json` in the [external secret](#manage-external-secret):

~~~ yaml
config.json: |
  {
    "SSL_CA": "/app/certs",
    "SSL_CERT": "/app/certs",

>     "SSL_KEY": "/app/certs",

Reviewer 2: The path needs to contain the file name and the end.

For example:

~~~ yaml
lms:
  sslVolumes:
    - name: source-ca
      secret:
        secretName: source-ca
    - name: source-cert
      secret:
        secretName: source-cert
    - name: source-key
      secret:
        secretName: source-key
    - name: cockroach-ca
      secret:
        secretName: cockroach-ca
    - name: cockroach-cert
      secret:
        secretName: cockroach-cert
    - name: cockroach-key
      secret:
        secretName: cockroach-key
  sslVolumeMounts:
    - mountPath: "/app/certs"
      name: source-ca
      readOnly: true
    - mountPath: "/app/certs"
      name: source-cert
      readOnly: true
    - mountPath: "/app/certs"
      name: source-key
      readOnly: true
    - mountPath: "/app/certs"
      name: cockroach-ca
      readOnly: true
    - mountPath: "/app/certs"
      name: cockroach-cert
      readOnly: true
    - mountPath: "/app/certs"
      name: cockroach-key
      readOnly: true
~~~

Specify the corresponding client certificates and keys in the [database connection strings](#source-and-target-connections). For example, within `config.json` in the [external secret](#manage-external-secret):

~~~ json
config.json: |
  {

>     "INIT_SOURCE": "mysql://{username}:{password}@{protocol}({host}:{port})/{database}?sslmode=verify-full?sslrootcert=path/to/mysql.ca&sslcert=path/to/mysql.crt&sslkey=path/to/mysql.key",

Reviewer 2: Same comment as above regarding the protocol bit in the mysql conn string.

    "INIT_TARGET": "postgresql://{username}:{password}@{host}:{port}/{database}?sslmode=verify-full?sslrootcert=path/to/ca.crt&sslcert=path/to/client.username.crt&sslkey=path/to/client.username.key"
  }
~~~

#### Configure orchestrator and client certificates

Mount the orchestrator certificate, key, and (optional) CA certificate to separate volumes. Cockroach Labs recommends mounting certificates to `/app/certs`.

For example:

> ~~~ yaml

Reviewer 3: Actually for the TLS setup for orchestrator, it will actually look something like this:
```
sslVolumes:
     - name: orch-tls
       secret:
         secretName: orch-tls
  sslVolumeMounts:
    - mountPath: "/app/certs"
      name: orch-tls
      readOnly: true

  env:
    - name: ORCH_CA_TLS_CERT
      value: /app/certs/ca.crt
    - name: ORCH_TLS_CERT
      value: /app/certs/tls.crt
    - name: ORCH_TLS_KEY
      value: /app/certs/tls.key
```
Follow this diff again: https://github.com/cockroachdb/molt/compare/main...user/rluu/run-lms-secrets-and-tls

This does depend on having a separate `orch-tls` external secret that looks like this:
```

apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: orch-tls
spec:
  refreshInterval: 1h
  secretStoreRef:
    name: aws-secret-store
    kind: SecretStore
  target:
    name: orch-tls
    creationPolicy: Owner
    template:
      engineVersion: v2
      data:
        ca.crt: '{{ .caCert }}'
        tls.crt: '{{ .serverCert }}'
        tls.key: '{{ .serverKey }}'
  data:
  - secretKey: caCert
    remoteRef:
      key: orch-certs
      property: caCert
  - secretKey: serverCert
    remoteRef:
      key: orch-certs
      property: serverCert
  - secretKey: serverKey
    remoteRef:
      key: orch-certs
      property: serverKey
```

Reviewer 3: As you can see in the example, you can do this through ENV variables as well. But the config .json you have below looks reasonable too.

orchestrator:
  sslVolumes:
    - name: orch-ca
      secret:
        secretName: orch-ca
    - name: orch-cert
      secret:
        secretName: orch-cert
    - name: orch-key
      secret:
        secretName: orch-key
  sslVolumeMounts:
    - mountPath: "/app/certs"
      name: orch-ca
      readOnly: true
    - mountPath: "/app/certs"
      name: orch-cert
      readOnly: true
    - mountPath: "/app/certs"
      name: orch-key
      readOnly: true
~~~

Set the [corresponding config keys](#cert-var) to the same mount path. For example, within `config.json` in the [external secret](#manage-external-secret):

~~~ yaml
config.json: |
  {
    "CA_TLS_CERT": "/app/certs",
    "TLS_CERT": "/app/certs",

>     "TLS_KEY": "/app/certs",

Reviewer 2: Same as above on the path needs to include the file name.

  }
~~~

This also requires that you create and specify a CLI client certificate, key, and (optional) CA certificate. It's easiest to specify these as environment variables in the shell that is running `molt-lms-cli`:

{% include_cached copy-clipboard.html %}
~~~ shell
export CLI_TLS_CA_CERT="{path-to-cli-ca-cert}"
~~~

{% include_cached copy-clipboard.html %}
~~~ shell
export CLI_TLS_CLIENT_CERT="{path-to-cli-client-cert}"
~~~

{% include_cached copy-clipboard.html %}
~~~ shell
export CLI_TLS_CLIENT_KEY="{path-to-cli-client-key}"
~~~

## `molt-lms-cli`

The `molt-lms-cli` command-line interface is used to inspect the LMS instances and [perform cutover](#perform-a-cutover).

> To install `molt-lms-cli`, [download the binary](https://molt.cockroachdb.com/lms/cli/).

Reviewer 4: This will have to be a direct link to the binaries. First we'll start with:
https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.darwin-arm64.tgz
https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.darwin-amd64.tgz

	To ensure data integrity, shadowing must be enabled for a sufficient duration with a low error rate. All LMS instances should have been continuously shadowing your workload for the past **seven days** at minimum, with only transient inconsistencies caused by events such as [transaction retry errors]({% link {{ page.version.version }}/transaction-retry-error-reference.md %}). The longer shadowing has been enabled, the better this allows you to evaluate consistency.

1. Once nearly all data from the source database is replicated to CockroachDB (for example, with a <1 second delay or <1000 rows), initiate the cutover:

	{% include_cached copy-clipboard.html %}
	~~~ shell
	molt-lms-cli cutover immediate {flags}
	~~~

	This command tells the LMS to switch the source of truth to CockroachDB. Application traffic is immediatedly directed to CockroachDB.

1. Any writes that were made during the cutover will have been missed on CockroachDB. Use [MOLT Verify]({% link {{ page.version.version }}/molt-verify.md %}) to identify the inconsistencies. These will need to be manually reconciled.
{% endcomment %}

> ## See also

Reviewer 3: @taroface @Jeremyyang920 are we not including instructions on how to helm install or upgrade this?

Reviewer 2: I don't know if we need to document how to use helm install as one of the dependencies of this is helm already. So you can look at the helm docs ryan already linked above

## Terminology

- A *live migration* keeps two production databases online (a source and a target database) and uses either replication or dual writing to keep data identical between them until a final cutover.
- The *source of truth* is the database that serves reads and writes to the application during a live migration. A cutover switches the source of truth.
- *Shadowing* is the replaying of traffic on the target database. The LMS supports multiple [shadowing modes](#shadowing-modes).

## Requirements

- [Kubernetes](https://kubernetes.io/) cluster
- [Helm](https://helm.sh/docs/intro/install/) package manager for Kubernetes

#### Supported databases

- [PostgreSQL]({% link {{ page.version.version }}/migrate-from-postgres.md %}) (source)
- [MySQL]({% link {{ page.version.version }}/migrate-from-mysql.md %}) (source)
- CockroachDB (source and target)

## Installation

Add the [Helm chart repository](https://molt.cockroachdb.com/lms/charts/) with [`helm repo add`](https://helm.sh/docs/helm/helm_repo_add/). Then install the chart with [`helm install`](https://helm.sh/docs/helm/helm_install/).

> To set up the LMS resources, [install `molt-lms-cli`](#molt-lms-cli) and run the following command:

Reviewer 3: When I click this, it takes me to this:
![image](https://github.com/cockroachdb/docs/assets/23587815/7f382bcf-b72b-48c8-860f-8fb7c0074ebd)

It shows only options for Intel and ARM. But what we really want to point out here are that there are multiple platforms and OSes that we support.

We support actually 6 combinations:
https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.windows-arm64.tgz
https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.windows-amd64.tgz[ https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.linux-arm64.tgz](https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.linux-arm64.tgz)[ https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.linux-amd64.tgz](https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.linux-amd64.tgz)[ https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.darwin-arm64.tgz](https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.darwin-arm64.tgz)[ https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.darwin-amd64.tgz](https://molt.cockroachdb.com/lms/cli/molt-lms-cli-0.1.0.darwin-amd64.tgz)

| `begin`              | Begin a consistent cutover. This pauses traffic to the source database.                                                                                                     |
| `commit`             | Commit a consistent cutover. This resumes traffic on the target database. This is only effective after running `cutover consistent begin`.                                  |
| `abort`              | Abort a consistent cutover after running `consistent cutover begin`, unless you have also run `consistent cutover commit`. This resumes traffic to the source database.     |
| `status`             | Display the current configuration of the LMS instances.                                                                                                                     |

{% comment %}
| `cutover immediate`  | Initiate an [immediate cutover](#immediate-cutover). This switches the source of truth to the target database. For usage details, see [Immediate cutover](#immediate-cutover). |
{% endcomment %}

> ### Flags

Reviewer 3: It might be worth mentioning here that for the orchestrator-url and TLS flags, you can supply the `CLI_ORCHESTRATOR_URL`, `CLI_TLS_CA_CERT`, etc. env vars and export them in the shell in order to not type them every time.

Author: Yes, this is already noted in the flag descriptions!

`shadowMode: sync` writes to both databases.

- The LMS sends application requests to the source of truth and the target database, and waits for each to respond.
- Query results from the source of truth are returned to the application.
- Query results from the non-source of truth are discarded.

{% comment %}
You can use this mode to perform an [immediate cutover](#immediate-cutover).
{% endcomment %}

### `strict-sync`

<img src="{{ 'images/v23.2/migrations/lms_strict-sync.svg' | relative_url }}" alt="MOLT LMS shadowing mode - strict-sync" />

`shadowMode: strict-sync` writes to both databases and enforces correctness on both databases.

- The LMS sends application requests to the source of truth and the target database, and waits for each to respond.
- Query results from the source of truth are returned to the application.
- If the non-source of truth returns an error, that error is returned instead of the query result from the source of truth.
- If both databases return an error, the non-source of truth will return the source-of-truth error.

{% comment %}
You can use this mode to perform an [immediate cutover](#immediate-cutover).
{% endcomment %}

> ## Perform a cutover

Reviewer 3: @ZhouXing19 do we support automatic consistent cutover here? If so, should we describe what that process entails?

Reviewer 5: We don't support automatic flow currently as we still miss this part: https://github.com/cockroachlabs/crdb-proxy/issues/539
This is the only missing part for automatic cutover, and It's on my plate for the Oct milestone

---
title: Use the Live Migration Service
summary: Learn how to use the Live Migration Service to shadow application traffic, perform cutover, and migrate a database to CockroachDB.
toc: true
docs_area: migrate
---

{{site.data.alerts.callout_info}}
{% include feature-phases/preview.md %}
{{site.data.alerts.end}}

MOLT LMS (Live Migration Service) is used to perform a [live migration]({% link {{ page.version.version }}/migration-overview.md %}#minimal-downtime) to CockroachDB.

> The LMS is a self-hosted, horizontally scalable proxy that routes traffic between an application, a source database, and a target CockroachDB database. You can use the LMS to control which database, as the "source of truth", is serving reads and writes to an application. You can optionally configure the LMS to [shadow production traffic](#shadowing-modes) from the source database and validate the query results on CockroachDB. When you have sufficiently tested your application and are confident with its consistency and performance on CockroachDB, you use the LMS to [perform the cutover](#perform-a-cutover) to CockroachDB.

Reviewer 6: ```suggestion
The LMS is a self-hosted, horizontally scalable proxy that routes traffic between an application, a source database, and a target CockroachDB database. You can use the LMS to control which database, as the "source of truth", is serving reads and writes to an application. You can optionally configure the LMS to [shadow production traffic](#shadowing-modes) from the source database and validate the query results on CockroachDB. When you have sufficiently tested your application and are confident with its consistency and performance on CockroachDB, you can use the LMS to [perform the cutover](#perform-a-cutover) to CockroachDB.
```

MOLT LMS is self-hosted on [Kubernetes](https://kubernetes.io/) and [configured using Helm](#configuration). At a high level, the LMS consists of the following:

- A number of proxy [instances](#lms-instances) (running in separate Kubernetes pods) across which application traffic is distributed and routed to the source and target databases.
- An "orchestrator" service (running in a single Kubernetes pod) that coordinates the proxy instances and sends the cutover commands.

This page describes how to [install](#installation), [configure](#configuration), [secure](#security), and [use the LMS](#molt-lms-cli) to perform a live migration. {% comment %}For more information, see [Migration Strategy: Live Migration]({% link {{ page.version.version }}/migration-strategy-live-migration.md %}).{% endcomment %}

## Terminology

> - A *live migration* keeps two production databases online (a source and a target database) and uses either replication or dual writing to keep data equal between them until a final cutover.

Reviewer 6: ```suggestion
- A *live migration* keeps two production databases online (a source and a target database) and uses either replication or dual writing to keep data identical between them until a final cutover.
```

~~~ yaml
lms:
  sslVolumes:
    - name: lms-tls
      secret:
        secretName: lms-tls
  sslVolumeMounts:
    - mountPath: "/app/certs"
      name: lms-tls
      readOnly: true
  env:
    - name: LMS_SSL_CA
      value: /app/certs/lms-ca.crt
    - name: LMS_SSL_CERT
      value: /app/certs/lms-tls.crt
    - name: LMS_SSL_KEY
      value: /app/certs/lms-tls.key
~~~

#### Configure orchestrator and client certificates

Create an external secret that specifies the orchestrator certificate, key, and (optional) CA certificate.

For example, the following `ExternalSecret` called `orch-tls` uses AWS Secrets Manager as the [`SecretStore`](https://external-secrets.io/latest/introduction/getting-started/#create-your-first-secretstore), and references a remote [AWS secret](https://docs.aws.amazon.com/secretsmanager/latest/userguide/create_secret.html) called `orch-certs`:

~~~ yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: orch-tls
spec:
  refreshInterval: 1h
  secretStoreRef:
    name: aws-secret-store
    kind: SecretStore
  target:
    name: orch-tls
    creationPolicy: Owner
    template:
      engineVersion: v2
      data:

>         orch-ca.crt: '{% raw %}{{ .caCert }}{% endraw %}'

Author: @Jeremyyang920 @ryanluu12345 I'm just guessing that these need to be the actual file names of the certs. Let me know if that's wrong.
